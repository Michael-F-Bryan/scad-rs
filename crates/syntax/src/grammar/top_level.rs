use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
