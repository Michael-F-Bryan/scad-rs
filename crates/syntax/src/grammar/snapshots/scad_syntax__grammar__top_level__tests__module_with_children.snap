---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(\n            translate([a, b, c]) {\n                cube(10);\n                cube(20);\n            })"
---
Node(
    MODULE_INSTANTIATION@0..101
      WHITESPACE@0..13 "\n            "
      IDENT@13..22 "translate"
      L_PAREN@22..23 "("
      ARGUMENTS@23..32
        LIST_EXPR@23..32
          L_BRACKET@23..24 "["
          LOOKUP_EXPR@24..25
            IDENT@24..25 "a"
          COMMA@25..26 ","
          LOOKUP_EXPR@26..28
            WHITESPACE@26..27 " "
            IDENT@27..28 "b"
          COMMA@28..29 ","
          LOOKUP_EXPR@29..31
            WHITESPACE@29..30 " "
            IDENT@30..31 "c"
          R_BRACKET@31..32 "]"
      R_PAREN@32..33 ")"
      BRACED_CHILDREN@33..101
        WHITESPACE@33..34 " "
        L_CURLY@34..35 "{"
        CHILDREN@35..87
          MODULE_INSTANTIATION@35..61
            WHITESPACE@35..52 "\n                "
            IDENT@52..56 "cube"
            L_PAREN@56..57 "("
            ARGUMENTS@57..59
              INTEGER@57..59
                INTEGER@57..59 "10"
            R_PAREN@59..60 ")"
            SEMICOLON@60..61
              SEMICOLON@60..61 ";"
          MODULE_INSTANTIATION@61..87
            WHITESPACE@61..78 "\n                "
            IDENT@78..82 "cube"
            L_PAREN@82..83 "("
            ARGUMENTS@83..85
              INTEGER@83..85
                INTEGER@83..85 "20"
            R_PAREN@85..86 ")"
            SEMICOLON@86..87
              SEMICOLON@86..87 ";"
        WHITESPACE@87..100 "\n            "
        R_CURLY@100..101 "}"
    ,
)
